File: Two_Pointers/11_Container_With_Most_Water/solution11.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int max_area = 0;
        int left = 0;
        int right = height.size() - 1;

        while (left < right) {
            int area = min(height[left], height[right]) * (right - left);
            max_area = max(max_area, area);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        
        return max_area;
    }
};

// Runtime1: 57ms, beats 76.79%
// Memory1: 61.63MB, beats 74.81%

// Runtime2: 52ms, beats 90.80%
// Memory2: 61.69MB, beats 74.81%

// Runtime3: 55ms, beats 84.07%
// Memory3: 61.48MB, beats 98.64%

File: Two_Pointers/11_Container_With_Most_Water/solution11.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_area = 0
        left = 0
        right = len(height) - 1

        while left < right:
            area = min(height[left], height[right]) * (right - left)
            max_area = max(max_area, area)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        return max_area

# Runtime1: 526ms, beats 51.90%
# Memory1: 29.87MB, beats 25.13%

# Runtime2: 532ms, beats 35.23%
# Memory2: 29.85MB, beats 25.13%

# Runtime3: 531ms, beats 37.75%
# Memory3: 29.88MB, beats 25.13%

# Runtime reduced by a factor of 10 to not skew data visualization